File: jvm/core/main/scala/com/webank/eggroll/core/resourcemanager/metadata/StoreCrudOperator.scala
/*
 * Copyright (c) 2019 - now, Eggroll Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *
 */

package com.webank.eggroll.core.resourcemanager.metadata

import java.util.concurrent.ConcurrentHashMap

import com.webank.eggroll.core.constant._
import com.webank.eggroll.core.error.CrudException
import com.webank.eggroll.core.meta._
import com.webank.eggroll.core.resourcemanager.ResourceDao
import com.webank.eggroll.core.util.JdbcTemplate.ResultSetIterator
import com.webank.eggroll.core.util.{Logging, TimeUtils}
import org.apache.commons.lang3.StringUtils

import scala.collection.mutable.ArrayBuffer

class StoreCrudOperator extends CrudOperator with Logging {

  def getOrCreateStore(input: ErStore): ErStore = synchronized {
    def doGetOrCreateStore(input: ErStore): ErStore = {
      val inputStoreLocator = input.storeLocator
      val inputWithoutType = input.copy(storeLocator = inputStoreLocator.copy(storeType = StringConstants.EMPTY))
      val inputStoreType = inputStoreLocator.storeType
      val existing = StoreCrudOperator.doGetStore(inputWithoutType)
      if (existing != null) {
        if (!existing.storeLocator.storeType.equals(inputStoreType)) {
          logWarning(
            s"store namespace: ${inputStoreLocator.namespace}, name: ${inputStoreLocator.name} " +
              s"already exist with store type: ${existing.storeLocator.storeType}. " +
              s"requires type: ${inputStoreLocator.storeType}")
        }
        existing
      } else {
        StoreCrudOperator.doCreateStore(input)
      }
    }

    doGetOrCreateStore(input)
  }

  def getStore(input: ErStore): ErStore = synchronized {
    StoreCrudOperator.doGetStore(input)
  }

  def deleteStore(input: ErStore): ErStore = synchronized {
    StoreCrudOperator.doDeleteStore(input)
  }

  def getStoreFromNamespace(input: ErStore): ErStoreList = {
    val storeWithLocatorOnly = StoreCrudOperator.getStoreLocators(input: ErStore)

    storeWithLocatorOnly.copy(stores = storeWithLocatorOnly.stores.map(s => StoreCrudOperator.doGetStore(s)))
  }
}

object StoreCrudOperator {
  private lazy val dbc = ResourceDao.dbc
  private val nodeIdToNode = new ConcurrentHashMap[java.lang.Long, DbServerNode]()
  private[metadata] def doGetStore(input: ErStore): ErStore = {
    val inputOptions = input.options

    // getting input locator
    val inputStoreLocator = input.storeLocator
    var queryStoreLocator = "select * from store_locator " +
      "where namespace = ? and name = ? and status = ?"
    var params = List(inputStoreLocator.namespace, inputStoreLocator.name, StoreStatus.NORMAL)

    if (!StringUtils.isBlank(inputStoreLocator.storeType)) {
      queryStoreLocator += " and store_type = ?"
      params ++= Array(inputStoreLocator.storeType)
    }

    queryStoreLocator += " limit 1"

    val storeLocatorResult = dbc.query(rs =>
      rs.map(_ => DbStoreLocator(
        id = rs.getLong("store_locator_id"),
        storeType = rs.getString("store_type"),
        namespace = rs.getString("namespace"),
        name = rs.getString("name"),
        path = rs.getString("path"),
        totalPartitions = rs.getInt("total_partitions"),
        partitioner = rs.getString("partitioner"),
        serdes = rs.getString("serdes")
      )), queryStoreLocator, params:_*).toList

    if (storeLocatorResult.isEmpty) {
      return null
    }

    val store = storeLocatorResult(0)
    val storeLocatorId = store.id

    val queryStorePartition = "select * from store_partition " +
      "where store_locator_id = ? order by store_partition_id asc"

    val storePartitionResult = dbc.query(rs => rs.map(_ => DbStorePartition(
      nodeId = rs.getLong("node_id"),
      partitionId = rs.getInt("partition_id"))), queryStorePartition, storeLocatorId).toList

    if (storePartitionResult.isEmpty) {
      throw new IllegalStateException("store locator found but no partition found")
    }

    val missingNodeId = ArrayBuffer[Long]()
    val partitionAtNodeIds = ArrayBuffer[Long]()

    for (i <- 0 until storePartitionResult.size){
      val nodeId = storePartitionResult(i).nodeId
      if (!nodeIdToNode.containsKey(nodeId)) missingNodeId += nodeId
      partitionAtNodeIds += nodeId
    }

    if (!missingNodeId.isEmpty) {
      var first = true
      val queryServerNode = new StringBuilder()
      queryServerNode.append(s"select * from server_node where status = '${ServerNodeStatus.HEALTHY}'")
        .append(s" and node_type = '${ServerNodeTypes.NODE_MANAGER}'")
        .append(s" and server_node_id in (")
      for(i <- 0 until missingNodeId.length){
        if (first) first = false else queryServerNode.append(", ")
        queryServerNode.append("?")
      }
      queryServerNode.append(")")

      val nodeResult = dbc.query(rs => rs.map(_ => DbServerNode(
        id = rs.getLong("server_node_id"),
        name = rs.getString("name"),
        clusterId = rs.getLong("server_cluster_id"),
        endpoint = ErEndpoint(host = rs.getString("host"), port = rs.getInt("port")),
        nodeType = rs.getString("node_type"),
        status = rs.getString("status"),
        lastHeartbeatAt = rs.getDate("last_heartbeat_at"),
        createdAt = rs.getDate("created_at"),
        updatedAt = rs.getDate("updated_at")
      )),
        queryServerNode.toString(),
        missingNodeId:_*).toList

      if (nodeResult.isEmpty) {
        throw new IllegalStateException(s"No valid node for this store: ${inputStoreLocator}")
      }

      for (i <- 0 until nodeResult.length){
        val serverNodeId = nodeResult(i).id
        nodeIdToNode.putIfAbsent(serverNodeId, nodeResult(i))
      }
    }

    val outputStoreLocator = ErStoreLocator(
      storeType = store.storeType,
      namespace = store.namespace,
      name = store.name,
      path = store.path,
      totalPartitions = store.totalPartitions,
      partitioner = store.partitioner,
      serdes = store.serdes)

    val storeOpts = dbc.query(
      rs => rs.map(
        _ => DbStoreOption(
          name = rs.getString("name"),
          data = rs.getString("data"),
          createdAt = rs.getDate("created_at"),
          updatedAt = rs.getDate("updated_at"))
      ),
      "select * from store_option where store_locator_id = ?", storeLocatorId)

    val outputOptions = new ConcurrentHashMap[String, String]()
    if (inputOptions != null) {
      outputOptions.putAll(inputOptions)
    }
    if (storeOpts != null) {
      storeOpts.foreach(r => outputOptions.put(r.name, r.data))
    }

    // process output partitions
    val outputPartitions = storePartitionResult.map(p => ErPartition(
        id = p.partitionId,
        storeLocator = outputStoreLocator,
        processor = ErProcessor(id = p.partitionId.toLong, serverNodeId = p.nodeId)))

    ErStore(storeLocator = outputStoreLocator, partitions = outputPartitions.toArray, options = outputOptions)
  }

  private[metadata] def doCreateStore(input: ErStore): ErStore = dbc.withTransaction(conn => {
    val inputOptions = input.options

    // create store locator
    val inputStoreLocator = input.storeLocator

    val sql = "insert into store_locator " +
      "(store_type, namespace, name, path, total_partitions, " +
      "partitioner, serdes, status) values (?, ?, ?, ?, ?, ?, ?, ?)"
    val newStoreLocator = dbc.update(conn, sql,
      inputStoreLocator.storeType,
      inputStoreLocator.namespace,
      inputStoreLocator.name,
      inputStoreLocator.path,
      inputStoreLocator.totalPartitions,
      inputStoreLocator.partitioner,
      inputStoreLocator.serdes,
      StoreStatus.NORMAL)

    if (newStoreLocator.isEmpty){
      throw new CrudException(s"Illegal rows affected returned when creating store locator: 0")
    }

    // create partitions
    var newTotalPartitions = inputStoreLocator.totalPartitions

    val newPartitions: ArrayBuffer[ErPartition] = ArrayBuffer[ErPartition]()
    newPartitions.sizeHint(inputStoreLocator.totalPartitions)

    val serverNodes: Array[ErServerNode] =
      ServerNodeCrudOperator.doGetServerNodes(
        input = ErServerNode(
          nodeType = ServerNodeTypes.NODE_MANAGER,
          status = ServerNodeStatus.HEALTHY)).sortBy(n => n.id)

    val nodesCount = serverNodes.length
    val specifiedPartitions = input.partitions
    val isPartitionsSpecified = specifiedPartitions.length > 0

    if (newTotalPartitions <= 0) newTotalPartitions = nodesCount << 2

    val serverNodeIds = ArrayBuffer[Long]()
    for (i <- 0 until newTotalPartitions) {

      val node: ErServerNode = serverNodes(i % nodesCount)

      val sql = "insert into store_partition (store_locator_id, node_id, partition_id, status) values (?, ?, ?, ?)"

      val nodeRecord = dbc.update(conn, sql,
        newStoreLocator.get,
        if (isPartitionsSpecified) input.partitions(i % specifiedPartitions.length).processor.serverNodeId else node.id,
        i,
        PartitionStatus.PRIMARY)

      if (nodeRecord.isEmpty) {
        throw new CrudException(s"Illegal rows affected when creating node: 0")
      }

      serverNodeIds += node.id
      newPartitions += ErPartition(
        id = i,
        storeLocator = inputStoreLocator,
        processor = ErProcessor(id = i,
          serverNodeId = if (isPartitionsSpecified) input.partitions(i % specifiedPartitions.length).processor.serverNodeId else node.id,
          tag = "binding"))
    }

    val newOptions = new ConcurrentHashMap[String, String]()
    if (inputOptions != null) newOptions.putAll(inputOptions)
    val itOptions = newOptions.entrySet().iterator()
    while(itOptions.hasNext){
      val entry = itOptions.next()
      dbc.update(conn,
        "insert into store_option(store_locator_id, name, data) values (?, ?, ?)",
        newStoreLocator.get,
        entry.getKey,
        entry.getValue)
    }

    val result = ErStore(
      storeLocator = inputStoreLocator,
      partitions = newPartitions.toArray,
      options = newOptions)

    result
  })

  private[metadata] def doDeleteStore(input: ErStore): ErStore = {
    val inputStoreLocator = input.storeLocator

    val outputStoreLocator = if (inputStoreLocator.name.equals("*")) {
      val result = dbc.withTransaction(conn => {
        var sql = "update store_locator set name = concat(name, ?), status = ? where namespace = ? and status = ? "
        val storeType = inputStoreLocator.storeType
        if (storeType.equals("*")) {
          dbc.update(conn, sql,
            s".${TimeUtils.getNowMs()}", StoreStatus.DELETED, inputStoreLocator.namespace, StoreStatus.NORMAL)
        } else {
          sql += "and store_type = ?"
          dbc.update(conn, sql,
            s".${TimeUtils.getNowMs()}", StoreStatus.DELETED, inputStoreLocator.namespace, StoreStatus.NORMAL, storeType)
        }
      })

      inputStoreLocator
    } else {
      val sql = "select * from store_locator " +
        "where store_type = ? and namespace = ? and name = ? and status = ? limit 1"

      val nodeResult = dbc.query(rs => rs.map(_ => DbStoreLocator(
        id = rs.getLong("store_locator_id"),
        name = rs.getString("name"))), sql,
        inputStoreLocator.storeType, inputStoreLocator.namespace,
        inputStoreLocator.name, StoreStatus.NORMAL).toList

      if (nodeResult.isEmpty) {
        return null
      }

      val nodeRecord = nodeResult(0)
      val nameNow = nodeRecord.name + "." + TimeUtils.getNowMs()

      val storeLocatorRecord = dbc.withTransaction(conn => {
        val sql = "update store_locator " +
          "set name = ?, status = ? where store_locator_id = ?"

        dbc.update(conn, sql, nameNow, StoreStatus.DELETED, nodeRecord.id)
      })

      inputStoreLocator.copy(name = nodeRecord.name)
    }
    ErStore(storeLocator = outputStoreLocator)
  }

  def getStoreLocators(input: ErStore): ErStoreList = {
    var sql = "select * from store_locator where status = 'NORMAL' and"
    val whereFragments = ArrayBuffer[String]()
    val args = ArrayBuffer[String]()

    val storeLocator = input.storeLocator
    val storeName = storeLocator.name
    val storeNamespace = storeLocator.namespace
    val storeType = storeLocator.storeType
    var storeNameNew = ""
    if (!StringUtils.isBlank(storeName)) {
      if (StringUtils.contains(storeName, "*")) {
        storeNameNew = storeName.replace('*', '%')
        args += storeNameNew
      } else {
        args += storeName
      }
      whereFragments += " name like ?"
    }

    if (!StringUtils.isBlank(storeNamespace)) {
      whereFragments += " namespace = ?"
      args += storeNamespace
    }

    sql += String.join(" and ", whereFragments: _*)

    dbc.query(rs => {
      val stores = ArrayBuffer[ErStore]()
      while (rs.next()) {

        stores += ErStore(
          storeLocator = ErStoreLocator(
            storeType = rs.getString("store_type"),
            name = rs.getString("name"),
            namespace = rs.getString("namespace"),
            totalPartitions = rs.getInt("total_partitions")
          ))
      }

      ErStoreList(stores = stores.toArray)
    }, sql, args: _*)
  }
}

File: jvm/cluster_manager/src/main/java/com/webank/eggroll/clustermanager/dao/impl/StoreCrudOperator.java
package com.webank.eggroll.clustermanager.dao.impl;

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.baomidou.mybatisplus.core.conditions.update.UpdateWrapper;
import com.eggroll.core.config.Dict;
import com.eggroll.core.exceptions.CrudException;
import com.eggroll.core.pojo.*;
import com.eggroll.core.utils.JsonUtil;
import com.webank.eggroll.clustermanager.entity.ServerNode;
import com.webank.eggroll.clustermanager.entity.StoreLocator;
import com.webank.eggroll.clustermanager.entity.StoreOption;
import com.webank.eggroll.clustermanager.entity.StorePartition;

import org.apache.commons.lang3.StringUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

@Service
public class StoreCrudOperator {

    @Autowired
    StoreLocatorService storeLocatorService;
    @Autowired
    StorePartitionService storePartitionService;
    @Autowired
    ServerNodeService serverNodeService;
    @Autowired
    StoreOptionService storeOptionService;

    private final Map<Long, Object> nodeIdToNode = new ConcurrentHashMap<>();
    private static final Logger LOGGER = LogManager.getLogger(StoreCrudOperator.class);

    private synchronized ErStore doGetStore(ErStore input) {
        Map<String, String> inputOptions = input.getOptions();

        // getting input locator
        ErStoreLocator inputStoreLocator = input.getStoreLocator();

        StoreLocator params = new StoreLocator();
        params.setNamespace(inputStoreLocator.getNamespace());
        params.setName(inputStoreLocator.getName());
        params.setStatus(Dict.NORMAL);

        if (!StringUtils.isBlank(inputStoreLocator.getStoreType())) {
            params.setStoreType(inputStoreLocator.getStoreType());
        }
        List<StoreLocator> storeLocatorResult = storeLocatorService.list(new QueryWrapper<>(params));
        if (storeLocatorResult.isEmpty()) {
            return null;
        }
        StoreLocator store = storeLocatorResult.get(0);
        Long storeLocatorId = store.getStoreLocatorId();

        QueryWrapper<StorePartition> queryStorePartition = new QueryWrapper<>();
        queryStorePartition.lambda().eq(StorePartition::getStoreLocatorId, storeLocatorId)
                .orderByAsc(StorePartition::getStorePartitionId);
        List<StorePartition> storePartitionResult = storePartitionService.list(queryStorePartition);
        if (storePartitionResult.isEmpty()) {
            throw new IllegalStateException("store locator found but no partition found");
        }

        List<Long> missingNodeId = new ArrayList<>();
        List<Long> partitionAtNodeIds = new ArrayList<>();

        for (StorePartition storePartition : storePartitionResult) {
            Long nodeId = storePartition.getNodeId();
            if (!nodeIdToNode.containsKey(nodeId)) missingNodeId.add(nodeId);
            partitionAtNodeIds.add(nodeId);
        }
        if (!missingNodeId.isEmpty()) {
            QueryWrapper<ServerNode> queryServerNode = new QueryWrapper<>();
            queryServerNode.lambda().eq(ServerNode::getStatus, Dict.HEALTHY)
                    .eq(ServerNode::getNodeType, Dict.NODE_MANAGER)
                    .in(ServerNode::getServerNodeId, missingNodeId);
            List<ServerNode> nodeResult = serverNodeService.list(queryServerNode);
            if (nodeResult.isEmpty()) {
                LOGGER.error("No valid node for this store: " + JsonUtil.object2Json(inputStoreLocator) + " missingId " + missingNodeId.toString());
                throw new IllegalStateException("No valid node for this store: " + JsonUtil.object2Json(inputStoreLocator));
            }

            for (ServerNode serverNode : nodeResult) {
                Long serverNodeId = serverNode.getServerNodeId();
                nodeIdToNode.putIfAbsent(serverNodeId, serverNode);
            }
        }

        ErStoreLocator outputStoreLocator = new ErStoreLocator(store.getStoreLocatorId()
                , store.getStoreType()
                , store.getNamespace()
                , store.getName()
                , store.getPath()
                , store.getTotalPartitions()
                , store.getPartitioner()
                , store.getSerdes());
        QueryWrapper<StoreOption> storeOptionWrapper = new QueryWrapper<>();
        storeOptionWrapper.lambda().eq(StoreOption::getStoreLocatorId, storeLocatorId);
        List<StoreOption> storeOpts = storeOptionService.list(storeOptionWrapper);
        ConcurrentHashMap<String, String> outputOptions = new ConcurrentHashMap<>();
        if (inputOptions != null) {
            outputOptions.putAll(inputOptions);
        }
        if (storeOpts != null) {
            for (StoreOption r : storeOpts) outputOptions.put(r.getName(), r.getData());
        }

        // process output partitions
        List<ErPartition> outputPartitions = new ArrayList<>();
        for (StorePartition p : storePartitionResult) {

            ErProcessor erProcessor = new ErProcessor();
            erProcessor.setId(p.getPartitionId());
            erProcessor.setServerNodeId(p.getNodeId());

            outputPartitions.add(new ErPartition(p.getPartitionId()
                    , outputStoreLocator
                    , erProcessor
                    , -1
            ));
        }
        return new ErStore(outputStoreLocator, outputPartitions, outputOptions);
    }

    private ErStore doCreateStore(ErStore input) {
        Map<String, String> inputOptions = input.getOptions();
        ErStoreLocator inputStoreLocator = input.getStoreLocator();

        StoreLocator newStoreLocator = new StoreLocator();
        newStoreLocator.setStoreType(inputStoreLocator.getStoreType());
        newStoreLocator.setNamespace(inputStoreLocator.getNamespace());
        newStoreLocator.setName(inputStoreLocator.getName());
        newStoreLocator.setPath(inputStoreLocator.getPath());
        newStoreLocator.setTotalPartitions(inputStoreLocator.getTotalPartitions());
        newStoreLocator.setPartitioner(inputStoreLocator.getPartitioner());
        newStoreLocator.setSerdes(inputStoreLocator.getSerdes());
        newStoreLocator.setStatus(Dict.NORMAL);
        boolean addStoreLocatorFlag = storeLocatorService.save(newStoreLocator);
        if (!addStoreLocatorFlag) {

            throw new CrudException("Illegal rows affected returned when creating store locator: 0");
        }

        int newTotalPartitions = inputStoreLocator.getTotalPartitions();
        List<ErPartition> newPartitions = new ArrayList<>();
        List<ErServerNode> serverNodes = serverNodeService.getListByErServerNode(
                new ErServerNode(
                        Dict.NODE_MANAGER
                        , Dict.HEALTHY)).stream().sorted(Comparator.comparingLong(ErServerNode::getId)).collect(Collectors.toList());
        int nodesCount = serverNodes.size();
        List<ErPartition> specifiedPartitions = input.getPartitions();
        boolean isPartitionsSpecified = specifiedPartitions.size() > 0;
        if (newTotalPartitions <= 0) newTotalPartitions = nodesCount << 2;
        ArrayList<Long> serverNodeIds = new ArrayList<>();
        for (int i = 0; i < newTotalPartitions; i++) {
            ErServerNode node = serverNodes.get(i % nodesCount);
            StorePartition nodeRecord = new StorePartition();
            nodeRecord.setStoreLocatorId(newStoreLocator.getStoreLocatorId());
            nodeRecord.setNodeId(isPartitionsSpecified ? input.getPartitions().get(i % specifiedPartitions.size()).getProcessor().getServerNodeId() : node.getId());
            nodeRecord.setPartitionId(i);
            nodeRecord.setStatus(Dict.PRIMARY);
            boolean addNodeRecordFlag = storePartitionService.save(nodeRecord);
            if (!addNodeRecordFlag) {

                throw new CrudException("Illegal rows affected when creating node: 0");
            }
            serverNodeIds.add(node.getId());

            ErProcessor binding = new ErProcessor();
            binding.setId(i);
            binding.setServerNodeId(isPartitionsSpecified ? input.getPartitions().get(i % specifiedPartitions.size()).getProcessor().getServerNodeId() : node.getId());
            binding.setTag("binding");

            newPartitions.add(new ErPartition(i
                    , inputStoreLocator
                    ,binding
                    , -1));
        }
        ConcurrentHashMap<String, String> newOptions = new ConcurrentHashMap<>();
        if (inputOptions != null) newOptions.putAll(inputOptions);
        for (Map.Entry<String, String> entry : newOptions.entrySet()) {
            StoreOption newStoreOption = new StoreOption();
            newStoreOption.setStoreLocatorId(newStoreLocator.getStoreLocatorId());
            newStoreOption.setName(entry.getKey());
            newStoreOption.setData(entry.getValue());
            storeOptionService.save(newStoreOption);
        }
        return new ErStore(inputStoreLocator, newPartitions, newOptions);
    }

    @Transactional
    private ErStore doDeleteStore(ErStore input) {
        ErStoreLocator inputStoreLocator = input.getStoreLocator();
        ErStoreLocator outputStoreLocator = null;
        if ("*" .equals(inputStoreLocator.getName())) {
            UpdateWrapper<StoreLocator> updateWrapper = new UpdateWrapper<>();
            updateWrapper.lambda().setSql("name = concat(name, " + System.currentTimeMillis() + ")")
                    .set(StoreLocator::getStatus, Dict.DELETED)
                    .eq(StoreLocator::getNamespace, inputStoreLocator.getNamespace())
                    .eq(StoreLocator::getStatus, Dict.NORMAL)
                    .eq("*" .equals(inputStoreLocator.getStoreType()), StoreLocator::getStoreType, inputStoreLocator.getStoreType());
            storeLocatorService.update(updateWrapper);
            outputStoreLocator = inputStoreLocator;
        } else {
            QueryWrapper<StoreLocator> queryWrapper = new QueryWrapper<>();
            queryWrapper.lambda().eq(StoreLocator::getStoreType, inputStoreLocator.getStoreType())
                    .eq(StoreLocator::getNamespace, inputStoreLocator.getNamespace())
                    .eq(StoreLocator::getName, inputStoreLocator.getName())
                    .eq(StoreLocator::getStatus, Dict.NORMAL);
            List<StoreLocator> nodeResult = storeLocatorService.list(queryWrapper);
            if (nodeResult.isEmpty()) {
                return null;
            }
            StoreLocator nodeRecord = nodeResult.get(0);
            String nameNow = nodeRecord.getName() + "." + System.currentTimeMillis();
            UpdateWrapper<StoreLocator> updateWrapper = new UpdateWrapper<>();
            updateWrapper.lambda().set(StoreLocator::getName, nameNow)
                    .set(StoreLocator::getStatus, Dict.DELETED)
                    .eq(StoreLocator::getStoreLocatorId, nodeRecord.getStoreLocatorId());
            storeLocatorService.update(updateWrapper);
            inputStoreLocator.setName(nameNow);
            outputStoreLocator = inputStoreLocator;
        }
        return new ErStore(outputStoreLocator, new ArrayList<>(), new ConcurrentHashMap<>());
    }

//    public ErStoreList getStoreLocators(ErStore input) {
//        QueryWrapper<StoreLocator> queryWrapper = new QueryWrapper<>();
//        ErStoreLocator storeLocator = input.getStoreLocator();
//        String storeName = storeLocator.getName();
//        String storeNamespace = storeLocator.getNamespace();
//        queryWrapper.apply(!StringUtils.isBlank(storeName), " and name like " + storeName.replace('*', '%'))
//                .lambda().eq(!StringUtils.isBlank(storeNamespace), StoreLocator::getNamespace, storeNamespace);
//        List<StoreLocator> storeList = storeLocatorService.list(queryWrapper);
//        List<ErStore> erStoreArr = new ArrayList<>();
//        for (int i = 0; i < storeList.size(); i++) {
//            StoreLocator store = storeList.get(i);
//            ErStoreLocator erStoreLocator = new ErStoreLocator(store.getStoreLocatorId()
//                    , store.getStoreType()
//                    , store.getNamespace()
//                    , store.getName()
//                    , store.getPath()
//                    , store.getTotalPartitions()
//                    , store.getPartitioner()
//                    , store.getSerdes());
//            erStoreArr.add(new ErStore(erStoreLocator, new ArrayList<>(), new ConcurrentHashMap<>()));
//        }
//        return new ErStoreList(erStoreArr, new ConcurrentHashMap<>());
//    }

}
